### ISSUE TITLE
testing: audit sample test coverage

### ISSUE THREAD

**Issue by tbpg** [GoogleCloudPlatform/golang-samples#1402]
Indirect function calls aren't covered by testing/sampletests. How often does that happen? For example, some tests use string-> function indirection.

**Comment by broady** [comment]
It's definitely used a bunch in vision tests I wrote.

```
$ git grep 'func(' | grep _test.go | grep -v testing.T | grep -v testutil.R| grep -v filepath.Walk | grep -v RunInTrans | grep -v '\:func'
bigtable/filters/filters_test.go:		filter func(io.Writer, string, string, string) error
bigtable/reads/reads_test.go:		filter func(io.Writer, string, string, string) error
container_registry/container_analysis/samples_test.go:		go func() {
dlp/snippets/jobs/jobs_test.go:	err = s.Receive(ctx, func(ctx context.Context, msg *pubsub.Message) {
firestore/firestore_snippets/save_test.go:	must := func(f func(context.Context, *firestore.Client) error) {
functions/http/request_test.go:	s := httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
functions/spanner/spanner_test.go:	defer func() {
internal/testutil/retry_test.go:	Retry(t, 5, time.Millisecond, func(r *R) {
internal/testutil/retry_test.go:	Retry(t, 10, time.Millisecond, func(r *R) {
iot/mqttsnippets/subscribe_gateway_device_topic_test.go:		go func() {
logging/simplelog/simplelog_test.go:	defer func() {
logging/simplelog/simplelog_test.go:	defer func() {
logging/simplelog/simplelog_test.go:	client.OnError = func(err error) {
pubsub/subscriptions/subscription_test.go:	once.Do(func() {
pubsub/topics/topics_test.go:	once.Do(func() {
regiontag_test.go:	go func() {
regiontag_test.go:	go func() {
regiontag_test.go:	output := func(c <-chan RegionTag) {
regiontag_test.go:	go func() {
run/authentication/auth_test.go:	metadata := httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
run/authentication/auth_test.go:	target := httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
run/grpc-ping/e2e_test.go:	resp, err := grpcRequest(pingURL.Host+":443", pingURL.String(), func(ctx context.Context, conn *grpc.ClientConn) (interface{}, error) {
run/grpc-ping/e2e_test.go:	resp, err = grpcRequest(pingURL.Host+":443", pingURL.String(), func(ctx context.Context, conn *grpc.ClientConn) (interface{}, error) {
run/hello-broken/main_test.go:	defer func() {
spanner/spanner_leaderboard/leaderboard_test.go:	defer func() {
spanner/spanner_snippets/snippet_test.go:	cleanup = func() {
spanner/spanner_snippets/snippet_test.go:	cleanup = func() {
videointelligence/video_analyze/video_analyze_test.go:		gcs         func(io.Writer, string) error
vision/detect/detect_test.go:		local, gcs  func(io.Writer, string) error
```

But my guess is that's all.

**Comment by tbpg** [comment]
There is a function call AST node in a `TestFoo` function that calls a function in the same package, where that function's declaration overlaps with a region tag.

There may be some cases where it's easier to update `sampletests` or the samples themselves.

**Comment by tbpg** [comment]
```
directory                tested	total	diff	pct
----------------------------------------------------
docs/                    0	142	142	0.00
spanner/                 0	64	64	0.00
firestore/               6	63	57	0.10
automl/                  0	55	55	0.00
datastore/               4	57	53	0.07
bigtable/                11	44	33	0.25
vision/                  26	53	27	0.49
healthcare/              41	61	20	0.67
appengine_flexible/      1	20	19	0.05
storage/                 62	80	18	0.78
iot/                     15	31	16	0.48
pubsub/                  23	32	9	0.72
logging/                 4	13	9	0.31
jobs/                    44	52	8	0.85
auth/                    0	8	8	0.00
videointelligence/       6	13	7	0.46
language/                5	12	7	0.42
endpoints/               0	7	7	0.00
bigquery/                81	88	7	0.92
memorystore/             0	6	6	0.00
translate/               16	21	5	0.76
profiler/                0	5	5	0.00
asset/                   3	8	5	0.38
speech/                  10	14	4	0.71
testing/                 4	7	3	0.57
opentelemetry/           0	3	3	0.00
monitoring/              32	35	3	0.91
dlp/                     22	25	3	0.88
trace/                   0	2	2	0.00
opencensus/              0	2	2	0.00
iap/                     0	2	2	0.00
iam/                     16	18	2	0.89
functions/               10	12	2	0.83
errorreporting/          0	2	2	0.00
container_registry/      11	13	2	0.85
texttospeech/            6	7	1	0.86
secretmanager/           14	15	1	0.93
kms/                     30	31	1	0.97
getting-started/         0	1	1	0.00
dialogflow/              6	7	1	0.86
dataproc/                2	3	1	0.67
tasks/                   2	2	0	1.00
securitycenter/          30	30	0	1.00
run/                     1	1	0	1.00
iotkit/                  0	0	0	NaN
internal/                0	0	0	NaN
container/               0	0	0	NaN
cloudsql/                0	0	0	NaN
cdn/                     2	2	0	1.00
appengine/               1	1	0	1.00
```

**Comment by tbpg** [comment]
I ran `sampletests` (modified for machine readable output) in each of the root directories. Here is the result.

**This only includes regions in the root module.** The other modules don't have very many samples (~100 total).

**Comment by tbpg** [comment]
1. `docs`: Almost all untested.
1. `spanner`: Almost all indirect function calls (`"awesomeSample" -> awesomeSample -> foo()`). Would be fixed by #1315.
1. `firestore`: Almost all untested.
1. `automl`: Untested.
1. `datastore`: Almost all untested.
1. `bigtable`: Mostly function references rather than function calls. Should be fixable in `sampletests` because it doesn't have the Spanner string double-indirection.
1. `vision`: Tests use function references. Same as `bigtable`.
1. `healthcare`: Many missing tests.
1. `appengine_flexible`: Untested.
1. `storage`: Many missing tests. Might be able to delete some if they moved to other files are are now duplicates.

That covers the top ten directories and a vast majority of the uncovered samples. I was hoping `sampletests` would be missing more samples. But, it looks like we're missing a lot of tests.

**Comment by tbpg** [comment]
The `spanner` sample rewrite is in progress.

#1406 mostly fixed `bigtable` and `vision`.

Other than that, the main issue is a legitimate lack of test coverage. So, I'm closing this.
